INWX/Domrobot.php

```diff
@@ -2,204 +2,223 @@
-			for ($j = 0; $j < 8; $j++) {
-				$x .= str_pad(base_convert(@$base32charsFlipped[@$secret[$i + $j]], 10, 2), 5, '0', STR_PAD_LEFT);
-			}
-			$eightBits = str_split($x, 8);
-			for ($z = 0; $z < count($eightBits); $z++) {
-				$binaryString .= ( ($y = chr(base_convert($eightBits[$z], 2, 10))) || ord($y) == 48 ) ? $y:"";
-			}
-		}
-		return $binaryString;
-	}
-	
-	private function _getBase32LookupTable() {
-		return array(
-				'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', // 7
-				'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', // 15
-				'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', // 23
-				'Y', 'Z', '2', '3', '4', '5', '6', '7', // 31
-				'=' // padding char
-		);
-	}
-}
-?>
+    }
+
+    private function _getSecretCode($secret)
+    {
+        $_timeSlice = floor(time() / 30);
+        $_codeLength = 6;
+
+        $secretKey = $this->_base32Decode($secret);
+        // Pack time into binary string
+        $time = chr(0) . chr(0) . chr(0) . chr(0) . pack('N*', $_timeSlice);
+        // Hash it with users secret key
+        $hm = hash_hmac('SHA1', $time, $secretKey, true);
+        // Use last nipple of result as index/offset
+        $offset = ord(substr($hm, -1)) & 0x0F;
+        // grab 4 bytes of the result
+        $hashPart = substr($hm, $offset, 4);
+
+        // Unpak binary value
+        $value = unpack('N', $hashPart);
+        $value = $value[1];
+        // Only 32 bits
+        $value = $value & 0x7FFFFFFF;
+
+        $modulo = pow(10, $_codeLength);
+        return str_pad($value % $modulo, $_codeLength, '0', STR_PAD_LEFT);
+    }
+
+    private function _base32Decode($secret)
+    {
+        if (empty($secret)) return '';
+
+        $base32chars = $this->_getBase32LookupTable();
+        $base32charsFlipped = array_flip($base32chars);
+
+        $paddingCharCount = substr_count($secret, $base32chars[32]);
+        $allowedValues = [6, 4, 3, 1, 0];
+        if (!in_array($paddingCharCount, $allowedValues)) return false;
+        for ($i = 0; $i < 4; $i++) {
+            if ($paddingCharCount == $allowedValues[$i] &&
+                substr($secret, -($allowedValues[$i])) != str_repeat($base32chars[32], $allowedValues[$i])) return false;
+        }
+        $secret = str_replace('=', '', $secret);
+        $secret = str_split($secret);
+        $binaryString = "";
+        for ($i = 0; $i < count($secret); $i = $i + 8) {
+            $x = "";
+            if (!in_array($secret[$i], $base32chars)) return false;
+            for ($j = 0; $j < 8; $j++) {
+                $x .= str_pad(base_convert(@$base32charsFlipped[@$secret[$i + $j]], 10, 2), 5, '0', STR_PAD_LEFT);
+            }
+            $eightBits = str_split($x, 8);
+            for ($z = 0; $z < count($eightBits); $z++) {
+                $binaryString .= (($y = chr(base_convert($eightBits[$z], 2, 10))) || ord($y) == 48) ? $y : "";
+            }
+        }
+        return $binaryString;
+    }
+
+    private function _getBase32LookupTable()
+    {
+        return [
```

Comment:
short-array-syntax is available since PHP 5.4. If people use older versions of PHP (which I don't hope! But happens), this will break their installation...
The version of this class/package should be increased to 3.0.0, and a CHANGELOG entry or warning should announce the new minimum-PHP-version. It also makes sense to put the PHP minimum version to composer.json